### PR TITLE
[Enhancement] Flush wal in parallel

### LINKED ISSUE #9878: Optimize meta data flush on BE when publish new version
## Enhancement

Currently we call `flushWAL(true /* sync */)` to ensure rowset meta data is persisted, but if we have multi disks, we need to flushWAL in serial order which have a big performance impact on version publish.

### PR DESCRIPTION
**Pull request by nshangyiming**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #9878

## Problem Summary(Required) ：

Using a thread pool to flush wal concurrently, after some simple tests we see the the flush time decrease by 80%~90%.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by mofeiatwork** [comment]
Any benchmark report for this PR?

**Comment by nshangyiming** [comment]
> Any benchmark report for this PR?

@mofeiatwork had a simple test with 3 BEs, eache BE has 12-14 disks, we can see the flush time decrease by 80%~90%.

File: be/src/storage/txn_manager.cpp [hunk 1]
```diff
@@ -49,6 +50,9 @@ TxnManager::TxnManager(int32_t txn_map_shard_size, int32_t txn_shard_size)
     _txn_map_locks = std::unique_ptr<std::shared_mutex[]>(new std::shared_mutex[_txn_map_shard_size]);
     _txn_tablet_maps = std::unique_ptr<txn_tablet_map_t[]>(new txn_tablet_map_t[_txn_map_shard_size]);
     _txn_partition_maps = std::unique_ptr<txn_partition_map_t[]>(new txn_partition_map_t[_txn_map_shard_size]);
+    auto st = ThreadPoolBuilder("wal-flush").set_min_threads(1).set_max_threads(32).set_idle_timeout(MonoDelta::FromSeconds(30))
```
[the remark this answers is not in the document]
Comment (nshangyiming) [reply be/src/storage/txn_manager.cpp]: changed to `meta-flush` to avoid misunderstand, and also max threads changed based on num of storage path.

**Review by chaoyli — DISMISSED** [review]

File: be/src/storage/txn_manager.cpp [hunk 2]
```diff
@@ -49,6 +50,12 @@ TxnManager::TxnManager(int32_t txn_map_shard_size, int32_t txn_shard_size)
     _txn_map_locks = std::unique_ptr<std::shared_mutex[]>(new std::shared_mutex[_txn_map_shard_size]);
     _txn_tablet_maps = std::unique_ptr<txn_tablet_map_t[]>(new txn_tablet_map_t[_txn_map_shard_size]);
     _txn_partition_maps = std::unique_ptr<txn_partition_map_t[]>(new txn_partition_map_t[_txn_map_shard_size]);
+    auto st = ThreadPoolBuilder("meta-flush")
```
Comment (chaoyli) [inline be/src/storage/txn_manager.cpp:53]: Use underscore instead of the horizontal line.
```suggestion
    auto st = ThreadPoolBuilder("meta_flush")
```

Comment (nshangyiming) [reply be/src/storage/txn_manager.cpp:53]: i think use '-' in thread name is more common.

**Comment by nshangyiming** [comment]
run starrocks_admit_test

File: be/src/storage/txn_manager.cpp [hunk 3]
```diff
@@ -269,12 +289,24 @@ Status TxnManager::persist_tablet_related_txns(const std::vector<TabletSharedPtr
...
+    std::vector<std::pair<Status, std::string>> pair_vec;
+    pair_vec.reserve(affected_dirs.size());
+    auto token = _thread_pool_flush->new_token(ThreadPool::ExecutionMode::CONCURRENT);
     for (auto dir : affected_dirs) {
-        auto st = dir->get_meta()->flush();
-        if (!st.ok()) {
-            LOG(WARNING) << "failed to flush tablet meta, dir:" << dir->path() << " res:" << st;
+        Status st;
+        token->submit_func([&]() { st = std::move(dir->get_meta()->flush()); });
+        pair_vec.push_back(std::make_pair(st, dir->path()));
```
Comment (decster) [inline be/src/storage/txn_manager.cpp:299]: looks like `st` is not thread safe, `st` may get invalided

Comment (nshangyiming) [reply be/src/storage/txn_manager.cpp:299]: Status's move and copy constructor is properly implemented, i think there is no thread safe problem, plz give a detailed description.

Comment (chaoyli) [reply be/src/storage/txn_manager.cpp:299]: Does the std::move() can guarantee the atomic operation?
```
    Status& operator=(Status&& s) noexcept {
        if (this != &s) {
            Status tmp(std::move(s));
            std::swap(this->_state, tmp._state);
        }
        return *this;
    }
```

File: be/src/storage/txn_manager.cpp [hunk 4]
```diff
@@ -269,12 +289,24 @@ Status TxnManager::persist_tablet_related_txns(const std::vector<TabletSharedPtr
...
+
+    std::vector<std::pair<Status, std::string>> pair_vec;
+    pair_vec.reserve(affected_dirs.size());
+    auto token = _thread_pool_flush->new_token(ThreadPool::ExecutionMode::CONCURRENT);
     for (auto dir : affected_dirs) {
-        auto st = dir->get_meta()->flush();
-        if (!st.ok()) {
-            LOG(WARNING) << "failed to flush tablet meta, dir:" << dir->path() << " res:" << st;
+        Status st;
+        token->submit_func([&]() { st = std::move(dir->get_meta()->flush()); });
```
Comment (chaoyli) [inline be/src/storage/txn_manager.cpp:298]: Does the queue will be fulled and submit failed?

Comment (decster) [reply be/src/storage/txn_manager.cpp:298]: `st` is temp variable, you cannot use it in lambda, it may already be invalid when lambda is using it.

Comment (nshangyiming) [reply be/src/storage/txn_manager.cpp]: fixed, plz see the new commit

File: be/src/storage/txn_manager.h [hunk 5]
```diff
@@ -180,6 +180,9 @@ class TxnManager {
 
     std::unique_ptr<std::shared_mutex[]> _txn_map_locks;
 
+    // Dynamic thread pool used to concurrently flush WAL to disk
+    std::unique_ptr<ThreadPool> _thread_pool_flush;
```
Comment (chaoyli) [inline be/src/storage/txn_manager.h:184]:
```suggestion
    std::unique_ptr<ThreadPool> _flush_thread_pool;
```

File: be/src/storage/txn_manager.cpp [hunk 6]
```diff
@@ -245,20 +252,35 @@ Status TxnManager::persist_tablet_related_txns(const std::vector<TabletSharedPtr
...
+        persisted.insert(path);
+    }
 
-        auto st = tablet->data_dir()->get_meta()->flush();
+    auto token = _thread_pool_flush->new_token(ThreadPool::ExecutionMode::CONCURRENT);
+    std::vector<std::pair<Status, int64_t>> pair_vec(to_flush_tablet.size());
+    int i = 0;
+    for (auto& tablet : to_flush_tablet) {
+        auto dir = tablet->data_dir();
+        token->submit_func([&pair_vec, dir, i]() { pair_vec[i].first = std::move(dir->get_meta()->flush()); });
```
Comment (chaoyli) [inline be/src/storage/txn_manager.cpp:272]: how about the queue is full and fail to submit_func?

Comment (nshangyiming) [reply be/src/storage/txn_manager.cpp:272]: The threadpool is created with an unlimited queue by default which is ok in this situation i think.

File: be/src/storage/txn_manager.h (hunk 5 above)
Comment (nshangyiming) [reply be/src/storage/txn_manager.h:184]: ok

**Review by chaoyli — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**